### ISSUE TITLE
Prompt coders to log in when they do significant work logged out

### ISSUE THREAD

**Issue by outoftime** [popcodeorg/popcode#1129]
While it’s fine to use Popcode logged out for quickly practicing coding techniques, we don’t want students doing a significant amount of work without it being saved. The always-visible “log in to save” prompt in the top bar is definitely an improvement, but I think we can be a bit more annoying about it.

In particular, I would propose a banner notification when the student has done a significant amount of work logged out. “A significant amount of work” might be measured in time since they started working, or number of discrete changes to the current project, or some combination of the two.

At such time as the wire is tripped for “a significant amount of work has been done”, we’ll want to display a banner notification (probably using a not-yet-existing “warning” yellow) that informs the user that their work won’t be saved and prompts them to log in. I think this idea is entirely stolen from CodePen.

I think it’s probably best to only display this notification once per session.

**Comment by outoftime** [comment]
OK @MaggieWalker I just rebased https://github.com/popcodeorg/popcode/pull/1584 so feel free to grab that branch. Also can you go to https://github.com/popcodeorg and accept the invite so I can assign this issue to you? Thanks!

**Comment by MaggieWalker** [comment]
Got it! @outoftime

**Comment by wilsonj806** [comment]
Makes sense, I think for my message I was going for a message with a similar vibe to how some MMO games warn you multiple times if your play session is really long, but keeping it short and sweet is probably better. And you're right on the banner part, not sure where I was going with it.

As for implementation, I need to familiarize myself a bit more with how Popcode's state is managed and learn Redux Sagas real fast, but I'm open to your initial thoughts on it.

**Comment by outoftime** [comment]
OK cool! I don’t want to be overly prescriptive about the implementation, so I’ll hold back my thoughts until you’ve formulated a proposal. Let me know if you get stuck though and I’m happy to jump in with suggestions 🙂

**Comment by wilsonj806** [comment]
Yep, so in terms of implementation, I'm thinking that I'd be coding it as a standalone React component that has a `useEffect` hook for dispatching the notification and renders `null`. My train of thought for that is that what we're trying to doesn't really feel like it fits in any of the existing components since we're scheduling a notification. The implementation will probably be similar to the Code Sandbox I linked above.

And then just a quick question, are you guys using the new `useDispatch`/ `useSelector` hooks for React Redux?

**Comment by outoftime** [comment]
Hey @wilsonj806 so that’s not a bad idea but Popcode’s architecture rests heavily on the Redux layer and React components are mostly just for presentational concerns. I know that’s become a bit unfashionable in the React community but it does work pretty well! Would you consider putting the business logic for this feature in a saga or a logic?

**Comment by wilsonj806** [comment]
@outoftime Hey Matt, I would definitely be down to do that, but I've been a bit busy recently so I'm not sure I'll get around to messing around with it in a reasonable amount of time.
